specs/ClearBrowsingData.md

```diff
@@ -0,0 +1,289 @@
+
+In this document we describe the updated API. We'd appreciate your feedback.
+
+
+# Description
+We expose browsing data clearing in two different asynchronous APIs:
+```IDL
+HRESULT ClearBrowsingData(
+      [in] uint64 dataKinds,
+      [in] ICoreWebView2ClearBrowsingDataCompletedHandler *handler); 
+
+HRESULT ClearBrowsingDataInTimeRange(
+      [in] uint64 dataKinds, 
+      [in] double startTime,
+      [in] double endTime, 
+      [in] ICoreWebView2ClearBrowsingDataCompletedHandler *handler);
+```
+The first method takes a uint64 parameter that consists of one or more COREWEBVIEW2_BROWSING_DATA_KIND passed in as well as a handler which will indicate if the proper data has been cleared successfully. The handler will respond with one of three results, which indicate that the method was successful, interrupted, or failed. This method clears the data for all time. 
+
+The second method takes the same parameters for the dataKinds and handler, as well as a start and end time in which the API should clear the corresponding data between. The double time parameters correspond to how many seconds since the UNIX epoch. 
+ 
+The browsing data kinds that are supported are listed below. These data kinds follow a hierarchical structure in which nested bullet points are included in their parent bullet point's data kind. 
+Ex: DOM storage is encompassed in site data which is encompassed in the profile data. Each of the following bullets correspond to a COREWEBVIEW2_BROWSING_DATA_KIND. 
+
+* Profile
+  * Site Data
+    * DOM Storage: App Cache, File Systems, Indexed DB, Local Storage, Web SQL, Cache 
+        Storage
+    * Cookies 
+  * HTTP Cache 
+  * Download History
+  * General Autofill 
+  * Password Autosave
+  * Browsing History
+  * Settings  
+
+# Examples
+
+## Win32 C++
+```cpp
+
+/// Function to clear the autofill data 
+bool EnvironmentComponent::ClearAutofillData() 
+{
+    wil::com_ptr<ICoreWebView2Environment5> environment;
+    webView->get_Environment(&environment);
+    uint64_t data_kinds = COREWEBVIEW2_BROWSING_DATA_KIND_GENERAL_AUTOFILL |
+            COREWEBVIEW2_BROWSING_DATA_KIND_PASSWORD_AUTOSAVE;
+    CHECK_FAILURE(environment->ClearBrowsingData(
+        data_kinds,
+        Callback<ICoreWebView2ClearBrowsingDataCompletedHandler>(
+            [this](HRESULT error, COREWEBVIEW2_CLEAR_BROWSING_DATA_RESULT_KIND result_kind)
+                -> HRESULT {
+                LPCWSTR result;
+                switch (result_kind)
+                {
+                case COREWEBVIEW2_CLEAR_BROWSING_DATA_RESULT_KIND_SUCCEEDED:
+                    result = L"Succeeded";
+                    break;
+                case COREWEBVIEW2_CLEAR_BROWSING_DATA_RESULT_KIND_INTERRUPTED:
+                    result = L"Interrupted";
+                    break;
+                case COREWEBVIEW2_CLEAR_BROWSING_DATA_RESULT_KIND_FAILED:
```

Comment:
We can distinguish between those with different HRESULTS / different exceptions. The Try* pattern is for if clearing data can regularly fail in a non-exceptional manner that the end dev needs to always check. Otherwise if failure is not expected to be a regular occurrence and so is exceptional we can use exceptions and not use the Try pattern with an explicit bool.